### PR TITLE
adds logic for retrieving and processing land attributes to RasterConditionFetcher and RasterConditionTreatmentEligibilitySelector

### PR DESCRIPTION
**Pull request by strangeties**
To RasterConditionFetcher, 1) adds logic for retrieving land attributes and 2) fixes pixel position bug
To RasterConditionTreatmentEligibilitySelector, adds logic for filtering stands based on land attributes
To ForsysGenerationRequestParams, adds parameters for filtering stands based on land attributes.

--------

The concrete examples below use the following parameter values:
StandEligibilityParams.filter_by_buildings = True 
StandEligibilityParams.filter_by_slope = True 
StandEligibilityParams.max_slope_in_percent_rise = 70.0 (corresponds with 35 degrees)
StandEligibilityParams.filter_by_road_proximity = True 
StandEligibilityParams.max_distance_from_road_in_meters = 1200.0 (advised to set to 60 m, but that's not testable for 300m data).

Example 1 failed to run with patchmax_proj_size=20.
```
ERROR: forsys.views.project area generation error: Error in search_func(cpp_graph = private$..update_edgelist(), net = private$..update_net(),  : 
  No patches possible
```

Both examples succeeded with patchmax_proj_size=2.

Example 1
image
image

Example 2
image
image

Example 3
```
ERROR: forsys.views.project area generation error: Error in makegraph(el[, c("from", "to", "dist_adj")]) : 
  Identified global objects via static code inspection (function (i); {; proj_obj <- NA; tryCatch({; patch <- build_func(i, cpp_graph, net, a_max, a_min,; c_max, c_min); ...; return(NA); }); }). NAs are not allowed in the graph
```

### CODE REVIEW

File: src/planscape/forsys/forsys_request_params.py [hunk 1]
```diff
@@ -157,6 +157,33 @@ def __init__(self, request: HttpRequest):
...
+    filter_by_road_proximity: bool 
+    max_distance_from_road_in_meters: float
+
+    def __init__(self):
+        self.filter_by_buildings = False
+        self.filter_by_slope = False 
+        # This corresponds with 35 degrees.
+        self.max_slope_in_percent_rise = 70.0
+        self.filter_by_road_proximity = False 
+        # Advised to set this to ~60 meters, but that's not testable for 300m
```
Comment (runforyourlives) [inline src/planscape/forsys/forsys_request_params.py:184]: add a TODO to change this after 30m data is available?

File: src/planscape/forsys/forsys_request_params.py [hunk 2]
```diff
@@ -157,6 +157,33 @@ def __init__(self, request: HttpRequest):
             DbRequestParams._URL_WRITE_TO_DB, True)
         self.scenario = get_scenario_by_id(
             self.user, self._URL_SCENARIO_ID, params)
+        
+
+# Parameters for deciding whether a stand is eligible for treatment.
+# TODO: make it possible to parse these parameters from URL and/or read them 
+# from the DB.
+class StandEligibilityParams:
+    BUILDINGS_KEY = "buildings"
```
Comment (runforyourlives) [inline src/planscape/forsys/forsys_request_params.py:166]: I'd add a note about these - these appear to be constants that correspond to the different options that StandEligibilityParams can hold; but the constants are to be used in the actual requests to forsys - these themselves are not the parameters.

File: src/planscape/forsys/raster_condition_fetcher.py [hunk 3]
```diff
@@ -51,87 +73,107 @@ class RasterConditionFetcher:
...
+                             conditions, attributes, raster_geo) -> dict[str, RasterPixelValues]:
+        raster_values = {}
+        self._fetch_condition_raster_values(
+            conditions, raster_geo, raster_values)
+        self._fetch_attribute_raster_values(
+            attributes, raster_geo, raster_values)
+        return raster_values
+
     # Fetches condition raster values for a given GEOSGeometry and emits it in
     # a {condition name: ConditionPixelValues} dictionary.
```
Comment (runforyourlives) [inline src/planscape/forsys/raster_condition_fetcher.py:116]: update to RasterPixelValues.

File: src/planscape/forsys/raster_condition_fetcher.py [hunk 4]
```diff
@@ -51,87 +73,107 @@ class RasterConditionFetcher:
...
             values = get_condition_values_from_raster(geo, c.raster_name)
             if values is None:
                 raise Exception(
                     "plan has no intersection with condition raster, %s" %
                     name)
-            conditions_to_raster_values[name] = values
-        return conditions_to_raster_values
+            raster_values[name] = values
+
+    def _fetch_attribute_raster_values(
```
Comment (runforyourlives) [inline src/planscape/forsys/raster_condition_fetcher.py:129]: Worth adding documentation similar to _fetch_raster_values

File: src/planscape/forsys/forsys_request_params.py (hunk 1 above)
Comment (strangeties) [reply src/planscape/forsys/forsys_request_params.py:184]: Done.

File: src/planscape/forsys/forsys_request_params.py (hunk 2 above)
Comment (strangeties) [reply src/planscape/forsys/forsys_request_params.py:166]: They're necessary in RasterConditionFetcher and RasterConditionTreatmentEligibilitySelector. 

They aren't used in the forsys input dataframe.

I could move them to ForsysGenerationInput - thinking it through now.

File: src/planscape/forsys/raster_condition_fetcher.py (hunk 3 above)
Comment (strangeties) [reply src/planscape/forsys/raster_condition_fetcher.py:116]: Done.
I've fixed this throughout the file (oops).

File: src/planscape/forsys/raster_condition_fetcher.py (hunk 4 above)
Comment (strangeties) [reply src/planscape/forsys/raster_condition_fetcher.py:129]: Done. Added documentation, and fixed function flow.

**Review by runforyourlives — APPROVED** [review]

File: src/planscape/forsys/forsys_request_params.py (hunk 2 above)
Comment (strangeties) [reply src/planscape/forsys/forsys_request_params.py]: fixed - I've moved them from StandEligibilityParams to ForsysGenerationInput.

**Comment by strangeties** [comment]
Thanks for reviewing!

**Pull request merged**